### PR TITLE
aggfuncs: implement avg(distinct) functions

### PR DESCRIPTION
**Pull request by zz-jason**
## What have you changed? (mandatory)

Implement `AVG` functions which handles the `distinct` property:
- use `map[types.MyDecimal]bool` to check whether a decimal values already exists
- use `map[float64]bool` for float64 values.

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

-unit test
-explain test

## Refer to a related PR or issue link (optional)

to #6952

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @jackysp PTAL

File: executor/aggfuncs/func_avg.go [hunk 1]
```diff
@@ -121,6 +121,68 @@ func (e *avgPartial4Decimal) UpdatePartialResult(sctx sessionctx.Context, rowsIn
 	return nil
 }
 
+type partialResult4AvgDistinctDecimal struct {
+	partialResult4AvgDecimal
+	exists map[types.MyDecimal]bool
```
Comment (winoros) [inline executor/aggfuncs/func_avg.go]: map[types.MyDecimal]struct{}?

File: executor/aggfuncs/func_avg.go [hunk 2]
```diff
@@ -205,3 +267,54 @@ func (e *avgPartial4Float64) UpdatePartialResult(sctx sessionctx.Context, rowsIn
 	}
 	return nil
 }
+
+type partialResult4AvgDistinctFloat64 struct {
+	partialResult4AvgFloat64
+	exists map[float64]bool
```
Comment (winoros) [inline executor/aggfuncs/func_avg.go]: ditto

Comment (zz-jason) [reply executor/aggfuncs/func_avg.go]: `map[types.MyDecimal]bool` is much convenient to use and will not introduce much memory overhead.

Comment (zz-jason) [reply executor/aggfuncs/func_avg.go]: Here is an experiment:
- `map[int]bool` with 10M elements: 315M 
- `map[int]struct{}` with 10M elements: 286M

## 1. memory usage of `map[int]bool`
```go
package main

import (
        "fmt"
        "runtime"
)

func main() {
        numElements := 10000000
        var m1, m2 runtime.MemStats

        runtime.ReadMemStats(&m1)
        a := make(map[int]bool)
        for i := 0; i < numElements; i++ {
                a[i] = true
        }
        runtime.ReadMemStats(&m2)
        fmt.Println("Bytes allocated for A:", m2.Alloc-m1.Alloc)
}
```

The result is:
```sh
[jianzhang.zj:/tmp]
➜ go build a.go

[jianzhang.zj:/tmp]
➜ ./a
Bytes allocated for A: 315457888
```

## 2. memory usage of `map[int]struct{}`
```go
package main

import (
        "fmt"
        "runtime"
)

func main() {
        numElements := 10000000
        var m1, m2 runtime.MemStats

        runtime.ReadMemStats(&m1)
        b := make(map[int]struct{})
        for i := 0; i < numElements; i++ {
                b[i] = struct{}{}
        }
        runtime.ReadMemStats(&m2)
        fmt.Println("Bytes allocated for B:", m2.Alloc-m1.Alloc)
}
```

the result is:
```sh
[jianzhang.zj:/tmp]
➜ go build b.go

[jianzhang.zj:/tmp]
➜ ./b
Bytes allocated for B: 286228080
```

**Pull request merged**